File: src/melody_sync/models/song.py
from dataclasses import dataclass
from pathlib import Path


@dataclass(slots=True)
class Song:
    """
    Representa uma música da biblioteca do Melody Sync.
    """

    # ==========================
    # Arquivo
    # ==========================

    path: Path
    size: int

    # ==========================
    # Metadados
    # ==========================

    title: str | None = None
    artist: str | None = None
    album: str | None = None

    # ==========================
    # Informações Técnicas
    # ==========================

    duration: float | None = None
    bitrate: int | None = None
    sample_rate: int | None = None
    channels: int | None = None
    codec: str | None = None

    @property
    def filename(self) -> str:
        """
        Retorna o nome do arquivo.
        """
        return self.path.name

    @property
    def extension(self) -> str:
        """
        Retorna a extensão do arquivo em letras minúsculas,
        sem o ponto inicial.
        """
        return self.path.suffix.lower().lstrip(".")

    @property
    def is_lossless(self) -> bool:
        """
        Retorna True se o formato for lossless.
        """
        return self.extension in {
            "flac",
            "wav",
            "aiff",
        }

    @property
    def directory(self) -> Path:
        """
        Retorna o diretório da música.
        """
        return self.path.parent

    @property
    def size_mb(self) -> float:
        """
        Retorna o tamanho em MB.
        """
        return self.size / (1024 * 1024)

    @property
    def size_gb(self) -> float:
        return self.size / (1024 ** 3)

    @property
    def duration_minutes(self) -> float:
        """
        Retorna a duração em minutos.
        """
        return self.duration / 60 if self.duration else 0

    @property
    def has_metadata(self) -> bool:
        """
        Indica se a música possui metadados básicos.
        """
        return bool(
            self.title and self.title.strip()
            and self.artist and self.artist.strip()
        )

File: melody-sync-core/src/main/kotlin/com/melodysync/model/Song.kt
package com.melodysync.model

import java.nio.file.Path

data class Song(
    val path: Path,
    val size: Long,
    val title: String? = null,
    val artist: String? = null,
    val album: String? = null,
    val duration: Double? = null,
    val bitrate: Int? = null,
    val sampleRate: Int? = null,
    val channels: Int? = null,
    val codec: String? = null,
) {
    val filename: String
        get() = path.fileName.toString()

    val extension: String
        get() {
            val name = path.fileName.toString()
            val dot = name.lastIndexOf('.')
            return if (dot > 0 && dot < name.length - 1) name.substring(dot + 1).lowercase() else ""
        }

    val isLossless: Boolean
        get() = extension in setOf("flac", "wav", "aiff")
    
    val directory: Path
        get() = path.parent

    val sizeMb: Double
        get() = size / (1024.0 * 1024.0)

    val sizeGb: Double
        get() = size / (1024.0 * 1024.0 * 1024.0)

    val durationMinutes: Double
        get() = (duration ?: 0.0) / 60.0

    val hasMetadata: Boolean
        get() = title != null && title.isNotBlank() && artist != null && artist.isNotBlank()
}